identity/src/main/java/com/android/identity/DeviceResponseParser.java

```diff
@@ -220,17 +174,42 @@ PublicKey parseIssuerSigned(
             builder.setIssuerSignedAuthenticated(issuerSignedAuthenticated);
             builder.setIssuerCertificateChain(issuerAuthorityCertChain);
 
-            DataItem mobileSecurityObjectBytes = Util.cborDecode(
-                    Util.coseSign1GetData(issuerAuthDataItem));
-            DataItem mobileSecurityObject = Util.cborExtractTaggedAndEncodedCbor(
-                    mobileSecurityObjectBytes);
+//            DataItem mobileSecurityObjectBytes = Util.cborDecode(
+//                    Util.coseSign1GetData(issuerAuthDataItem));
+//            DataItem mobileSecurityObject = Util.cborExtractTaggedAndEncodedCbor(
+//                    mobileSecurityObjectBytes);
+//            byte[] encodedMobileSecurityObject = Util.cborEncode(mobileSecurityObject);
+            byte[] encodedMobileSecurityObject = Util.cborExtractTaggedCbor(Util.coseSign1GetData(issuerAuthDataItem));
+            MobileSecurityObjectParser.MobileSecurityObject parsedMso = new MobileSecurityObjectParser()
+                    .setMobileSecurityObject(encodedMobileSecurityObject).parse();
+
+            builder.setValidityInfoSigned(parsedMso.getSigned());
+            builder.setValidityInfoValidFrom(parsedMso.getValidFrom());
+            builder.setValidityInfoValidUntil(parsedMso.getValidUntil());
+            if (parsedMso.getExpectedUpdate() != null) {
+                builder.setValidityInfoExpectedUpdate(parsedMso.getExpectedUpdate());
+            }
 
-            Pair<PublicKey, Map<String, Map<Long, byte[]>>> msoResult =
-                    parseMso(mobileSecurityObject, expectedDocType);
-            final PublicKey deviceKey = msoResult.first;
-            Map<String, Map<Long, byte[]>> digestMapping = msoResult.second;
+            /* don't care about version for now */
+            String digestAlgorithm = parsedMso.getDigestAlgorithm();
+            if (!digestAlgorithm.equals("SHA-256")) {
+                throw new IllegalArgumentException("Unsupported digestAlgorithm '"
+                        + digestAlgorithm + "' in MSO");
+            }
```

Comment:
I think our tests now exercise all three supported digests so this would fail, no?

Reply:
the tests which support all three digests are in the `MobileSecurityObjectGeneratorTest`, so it doesn't use the `DeviceResponseParser`. I modified the check here (by removing it, knowing it gets checked against all 3 in the MSO parser call) and abstracted the digester to be able to take in the digestAlgorithm from the MSO. However, I'm not sure how to test the call paths for other digests in the `DeviceResponseGeneratorTest` since it seems to be determined by `IdentityCredentialStore.CIPHERSUITE_ECDHE_HKDF_ECDSA_WITH_AES_256_GCM_SHA256` which doesn't have other cipher suites to choose from

Reply:
I see, yea, agree removing the check here since we already check for that in `MobileSecurityObjectParser`. And, yes, I was wrong, this check wouldn't be exercised and that's fine. (Btw, it doesn't have anything to do with `cipherSuite` but that's a different story.)